File: packages/appmetrica_push_ios/ios/Classes/LateInvoker.swift
typealias InvokerFunction = () -> Void

public class LateInvoker: NSObject {
    static let shared = LateInvoker()
    
    private var isInvoked = false
    private var functions: [InvokerFunction] = []
    
    func invoke(fn: @escaping InvokerFunction) {
        isInvoked ? fn() : functions.append(fn)
    }
    
    func invokeAll() {
        isInvoked = true
        
        functions.forEach { fn in fn() }
        functions.removeAll()
    }
    
    func dispose() {
        functions.removeAll()
        
        isInvoked = false
    }
}

File: packages/appmetrica_push_huawei/android/src/main/kotlin/ru/madbrains/appmetrica_push_huawei/LateInvoker.kt
package ru.madbrains.appmetrica_push_huawei

object LateInvoker {
    private val list: MutableList<() -> Unit> = mutableListOf()
    private var isInvoked = false

    fun invoke(fn: () -> Unit) {
        if (isInvoked) {
            fn()
        } else {
            list.add(fn)
        }
    }

    fun invokeAll() {
        isInvoked = true
        for (fn in list) {
            fn()
        }
    }

    fun dispose() {
        list.clear()
        isInvoked = false
    }
}